packages/mui-material/src/Button/Button.js

```diff
@@ -318,7 +318,7 @@ const Button = React.forwardRef(function Button(inProps, ref) {
   const fullWidth = fullWidthProp || fullWidthContext || false;
   const size = sizeProp || sizeContext || 'medium';
   const variant = variantProp || variantContext || 'text';
-  const disableRipple = disableRippleProp || disableRippleContext || false;
+  const disableRipple = disableRippleContext || disableRippleProp;
```

Comment:
This may look dubious. But if I instead do `disableRippleProp || disableRippleContext`, the individual buttons having `disableRipple={false}` do not take priority over theme's `defaultProps.disableRipple`. `false || undefined` equals `undefined` and  `defaultProps.disableRipple` still wins (Reason to add the second test case).

And if we do `disableRippleProp ?? disableRippleContext`, the `ButtonGroup` behaviour will change as discussed in https://github.com/mui-org/material-ui/pull/28645#discussion_r738967535 and would be a breaking change.

This change ensures both `ButtonGroup` and theme's `defaultProps.disableRipple` works as expected like _before_ v5.1.0 along with keeping the use of `React.Context` API.
I know this is confusing, another better solution without any breaking changes would be appreciated.

Reply:
Isn't `disableFocusRipple` suffering from the same issue?

Reply:
> Aren't disableFocusRipple, disableElevation, disabled, fullWidth, etc. suffering from the same kind of priority issues?

Not all of them has issues.
`disableRipple` is the only regression from v5.0.6 to 5.1.0.
 `disabled`  does not work since long with theme's defaultProps from `MuiButtonBase`.
Others work.

> I don't feel like we took my comment in #28645 (comment) into account.

We did not consider it as more bugs but a change in behaviour.

Reply:
@ZeeshanTamboli can we add tests for all of these props and make sure they work the same?

Reply:
An example that demonstrates that 1. the proposed fix is not working 2. the other props are impacted https://codesandbox.io/s/basicbuttongroup-material-demo-forked-65rk8?file=/demo.js. 

Button 2 should have a ripple and Button 3 should not be full width.